pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,533 @@
+					return
+				}
+				c.recorder.Eventf(vm, k8score.EventTypeNormal, SuccessfulCreateVirtualMachineReason, "Created virtual machine: %v", vm.ObjectMeta.Name)
+			}()
+		}
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		// Only return the first error which occurred, the others will most likely be equal errors
+		return err
+	default:
+	}
+	return nil
+}
+
+// filterActiveVMs takes a list of VMs and returns all VMs which are not in a final state
+// Note that vms which have a deletion timestamp set, are still treated as active.
+// This is a difference to Pod ReplicaSets
+func (c *VMReplicaSet) filterActiveVMs(vms []virtv1.VirtualMachine) []virtv1.VirtualMachine {
+	filtered := []virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if !vm.IsFinal() {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// filterMatchingVMs takes a selector and a list of VMs. If the VM labels match the selector it is added to the filtered collection.
+// Returns the list of all VMs which match the selector
+func (c *VMReplicaSet) filterMatchingVMs(selector labels.Selector, vms []virtv1.VirtualMachine) []virtv1.VirtualMachine {
+	//TODO take owner reference into account
+	filtered := []virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if selector.Matches(labels.Set(vm.ObjectMeta.Labels)) {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// listVMsFromNamespace takes a namespace and returns all VMs from the VM cache which run in this namespace
+func (c *VMReplicaSet) listVMsFromNamespace(namespace string) ([]virtv1.VirtualMachine, error) {
+	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vms := []virtv1.VirtualMachine{}
+	for _, obj := range objs {
+		vms = append(vms, *obj.(*virtv1.VirtualMachine))
+	}
+	return vms, nil
+}
+
+// listControllerFromNamespace takes a namespace and returns all VMReplicaSets from the ReplicaSet cache which run in this namespace
+func (c *VMReplicaSet) listControllerFromNamespace(namespace string) ([]virtv1.VirtualMachineReplicaSet, error) {
+	objs, err := c.vmRSInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	replicaSets := []virtv1.VirtualMachineReplicaSet{}
+	for _, obj := range objs {
+		rs := obj.(*virtv1.VirtualMachineReplicaSet)
+		replicaSets = append(replicaSets, *rs)
+	}
+	return replicaSets, nil
+}
+
+// getMatchingController returns the first VMReplicaSet which matches the labels of the VM from the listener cache.
+// If there are no matching controllers, a NotFound error is returned.
+func (c *VMReplicaSet) getMatchingController(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachineReplicaSet, error) {
+	log := logging.DefaultLogger()
+	controllers, err := c.listControllerFromNamespace(vm.ObjectMeta.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO check owner reference, if we have an existing controller which owns this one
```

Comment:
can we do that now? or are we having to wait for something to get updated in client-go/apimachinery first?

Reply:
I would have to copy over the controller_ref_manager.go from k8s right now. It seems to be already perfectly well isolated there, but not yet in one of these libraries. 

We need a solution for all our controllers (vm, migration and this one) which follows the controller-ref proposal. It might be worth a separate PR. For as long as the user does not make labeling errors everything will work.

Reply:
I see.  We'll want to follow up on this